ssmconfig.go

```diff
@@ -48,59 +54,56 @@ type Provider struct {
 // The behavior of using the `default` and `required` tags on the same struct field is
 // currently undefined.
 func (p *Provider) Process(configPath string, c interface{}) error {
-
 	v := reflect.ValueOf(c)
 	if v.Kind() != reflect.Ptr || v.IsNil() {
 		return errors.New("ssmconfig: c must be a pointer to a struct")
 	}
-
 	v = reflect.Indirect(reflect.ValueOf(c))
 	if v.Kind() != reflect.Struct {
 		return errors.New("ssmconfig: c must be a pointer to a struct")
 	}
+	// get params required from struct
+	spec := make(map[string]fieldSpec)
+	buildStructSpec(configPath, v.Type(), &spec)
 
-	spec := buildStructSpec(configPath, v.Type())
-
+	// get params from ssm parameter store
 	params, invalidPrams, err := p.getParameters(spec)
 	if err != nil {
 		return errors.Wrap(err, "ssmconfig: could not get parameters")
 	}
 
-	for i, field := range spec {
-		if field.name == "" && field.defaultValue == "" {
-			continue
-		}
-
-		if _, ok := invalidPrams[field.name]; ok && field.required {
-			return errors.Errorf("ssmconfig: %s is required", invalidPrams[field.name])
-		}
+	// set values in struct
+	return setValues(v, params, invalidPrams, spec)
+}
 
-		value, ok := params[field.name]
-		if !ok {
-			value = field.defaultValue
+func buildStructSpec(configPath string, t reflect.Type, spec *map[string]fieldSpec) {
```

Comment:
Is there any reason this returns a pointer to a map? A map is already a pointer-ish type

```suggestion
func buildStructSpec(configPath string, t reflect.Type, spec map[string]fieldSpec) {
```